packages/dds/tree/src/feature-libraries/treeTextCursor.ts

```diff
@@ -51,113 +48,90 @@ import { fail } from "../util";
+     * When -1, get siblings is a FieldKey.
      */
-    public isRooted(): boolean {
-        return this.keyStack.length === this.siblingStack.length + 1;
-    }
+    protected index: number;
 
-    protected getNode(): JsonableTree {
-        return this.siblings[this.index];
+    /**
+     * Might start at special root where fields are detached sequences.
+     */
+    public constructor(root: JsonableTree) {
+        this.siblings = [root];
+        this.index = 0;
     }
 
-    get value(): Value {
-        return this.getNode().value;
+    public getCurrentFieldKey(): FieldKey {
+        // assert(this.mode === CursorLocationType.Fields, "must be in fields mode");
+        return getFieldKey(this.index, this.siblings);
     }
 
-    get type(): TreeType {
-        return this.getNode().type;
+    private getStackedFieldKey(height: number): FieldKey {
+        assert(height % 2 === 1, "must field height");
+        return getFieldKey(this.indexStack[height], this.siblingStack[height]);
     }
 
-    get keys(): Iterable<FieldKey> {
-        return Object.getOwnPropertyNames(getGenericTreeFieldMap(this.getNode(), false)) as Iterable<FieldKey>;
+    private getStackedNodeIndex(height: number): number {
+        assert(height % 2 === 0, "must be node height");
+        return this.indexStack[height];
     }
 
-    down(key: FieldKey, index: number): SynchronousNavigationResult {
-        const siblings = getGenericTreeField(this.getNode(), key, false);
-        const child = siblings[index];
-        if (child !== undefined) {
-            this.indexStack.push(this.index);
-            this.siblingStack.push(this.siblings);
-            this.keyStack.push(key);
-            this.siblings = siblings;
-            this.index = index;
-            return TreeNavigationResult.Ok;
-        }
-        return TreeNavigationResult.NotFound;
+    private getStackedNode(height: number): JsonableTree {
+        const index = this.getStackedNodeIndex(height);
+        return (this.siblingStack[height] as readonly JsonableTree[])[index];
     }
 
-    seek(offset: number): SynchronousNavigationResult {
-        const index = offset + this.index;
-        const child = this.siblings[index];
-        if (child !== undefined) {
-            this.index = index;
-            return TreeNavigationResult.Ok;
-        }
-        return TreeNavigationResult.NotFound;
+    public getCurrentFieldLength(): number {
+        // assert(this.mode === CursorLocationType.Fields, "must be in fields mode");
+        return this.getField().length;
     }
 
-    up(): SynchronousNavigationResult {
-        const index = this.indexStack.pop();
-        if (index === undefined) {
-            // At root already (and made no changes to current location)
-            return TreeNavigationResult.NotFound;
-        }
-
+    public enterChildNode(index: number): void {
+        // assert(this.mode === CursorLocationType.Fields, "must be in fields mode");
```

Comment:
This is possible. Merging these multi use functions looks like a good idea just considering this particular cursor implementation, but I suspect other implementations will have different behavior and thus combining them could have a perf cost.

Since code should always statically know which it should be doing, and merging them doesn't make any usage patterns we I thought of more concise (unlike with seek and nextField) I kept them separate. This also allows for an implementation to do more validation (though I disabled those assets here for perf reasons, some implementations might have cheaper asserts or actual debug builds, ex a wasm version).

We should keep an eye out for any cases where merging these would be really helpful though.

Another very different option instead of combining these is to make upToNode/field vboth valid at all times and go all the way up to the next node/field (possible 2 layers, ex: when at a node and calling upToNode). Combining them would remove that option.

For now I think I'll leave it as is, but we should keep the idea in mind (there is a comment on the API for (upToField) mentioning this.

Reply:
Yeah, I also prefer to have the caller explicitly invoke the appropriate method.  My rationale was that when the caller knows multiple steps on the path (as in the latter case) there might be a minor win.

(It could also easily be a loss if.  Just thought it was worth mentioning.)